### PR TITLE
Stream - Collectors

### PR DESCRIPTION
**Pull request by webmuniz**

### CODE REVIEW

File: src/academy/devdojo/maratonajava/javacore/streams/domain/LightNovel.java [hunk 1]
```diff
@@ -5,17 +5,25 @@
...
     private String title;
     private double price;
+    private Category category;
 
     public LightNovel(String title, double price) {
         this.title = title;
         this.price = price;
     }
 
+    public LightNovel(String title, double price, Category category) {
```
Comment (bryan-cda) [inline src/academy/devdojo/maratonajava/javacore/streams/domain/LightNovel.java:15]: You can also use **@AllArgsConstructor** from the Lombok project.

File: src/academy/devdojo/maratonajava/javacore/streams/domain/LightNovel.java [hunk 2]
```diff
@@ -5,17 +5,25 @@
...
+        this.title = title;
+        this.price = price;
+        this.category = category;
+    }
+
     @Override
     public String toString() {
```
Comment (bryan-cda) [inline src/academy/devdojo/maratonajava/javacore/streams/domain/LightNovel.java:22]: You can also use **@ToString** from the Lombok project.

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest01.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    ));
+
+    public static void main(String[] args) {
+        System.out.println(lightNovels.stream().count()); //count
+        System.out.println(lightNovels.stream().collect(Collectors.counting())); //count using Collectors
+
+        System.out.println("------------");
+
+        lightNovels.stream().max(Comparator.comparing(LightNovel::getPrice)).ifPresent(System.out::println); //Max price
+        lightNovels.stream().collect(Collectors.maxBy(Comparator.comparing(LightNovel::getPrice))).ifPresent(System.out::println); //Max price with Collectors
```
Comment (bryan-cda) [inline src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest01.java:31]: Can you scanner this Class with SonarLint?

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest02.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+        categoryLightNovelMap.put(Category.DRAMA, drama);
+        categoryLightNovelMap.put(Category.FANTASY, fantasy);
+        categoryLightNovelMap.put(Category.ROMANCE, romance);
+
+        System.out.println(categoryLightNovelMap);
+
+
+        //SAME FUNCTION USING STREAM:
+
```
Comment (bryan-cda) [inline src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest02.java:49]: Can you remove the blank line?

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest03.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        //GROUP BY PROMOTION AND CATEGORY
+        Map<Category, Map<Promotion, List<LightNovel>>> collect1 = lightNovels
+                .stream()
+                .collect(groupingBy(
+                LightNovel::getCategory,
+                groupingBy(ln -> (ln.getPrice() < 5) ? UNDER_PROMOTION : NORMAL_PRICE))
+                );
+
+        System.out.println(collect1);
+
```
Comment (bryan-cda) [inline src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest03.java:46]: Can you remove the blank line?

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest04.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            new LightNovel("Monogatari", 4.99, Category.DRAMA),
+            new LightNovel("Overlord", 8.99, Category.FANTASY)
+    ));
+
+    public static void main(String[] args) {
+
+        //COUNT BY CATEGORY
+        final Map<Category, Long> collect = lightNovels.stream().collect(groupingBy(LightNovel::getCategory, Collectors.counting()));
+        System.out.println(collect);
+
```
Comment (bryan-cda) [inline src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest04.java:31]: What is the pattern of case used on your comments? UPPER, lower or Pascal Case?

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest05.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            new LightNovel("Demon Slayer", 5.89, Category.ROMANCE),
+            new LightNovel("Demon Slayer", 5.89, Category.ROMANCE),
+            new LightNovel("Pokemon", 6.50, Category.DRAMA),
+            new LightNovel("Fullmetal alchemist", 5.99, Category.FANTASY),
+            new LightNovel("Monogatari", 4.99, Category.DRAMA),
+            new LightNovel("Overlord", 8.99, Category.FANTASY)
+    ));
+
+    public static void main(String[] args) {
+
```
Comment (bryan-cda) [inline src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest05.java:26]: Can you remove the blank line?

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest02.java (hunk 4 above)
Comment (webmuniz) [reply src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest02.java:49]: removed

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest03.java (hunk 5 above)
Comment (webmuniz) [reply src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest03.java:46]: removed

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest04.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+            new LightNovel("Demon Slayer", 5.89, Category.ROMANCE),
+            new LightNovel("Demon Slayer", 5.89, Category.ROMANCE),
+            new LightNovel("Pokemon", 6.50, Category.DRAMA),
+            new LightNovel("Fullmetal alchemist", 5.99, Category.FANTASY),
+            new LightNovel("Monogatari", 4.99, Category.DRAMA),
+            new LightNovel("Overlord", 8.99, Category.FANTASY)
+    ));
+
+    public static void main(String[] args) {
+
```
[the remark this answers is not in the document]
Comment (webmuniz) [reply src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest04.java:27]: removed

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest05.java (hunk 7 above)
Comment (webmuniz) [reply src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest05.java:26]: removed

File: src/academy/devdojo/maratonajava/javacore/streams/domain/LightNovel.java (hunk 1 above)
Comment (webmuniz) [reply src/academy/devdojo/maratonajava/javacore/streams/domain/LightNovel.java:15]: Yes, but in this case I need to follow the course structure.

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest01.java (hunk 3 above)
Comment (webmuniz) [reply src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest01.java:31]: Yes, but in this case I need to follow the course structure

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest02.java (hunk 4 above)
Comment (webmuniz) [reply src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest02.java:49]: removed

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest03.java (hunk 5 above)
Comment (webmuniz) [reply src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest03.java:46]: removed

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest04.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (webmuniz) [reply src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest04.java:27]: removed

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest04.java (hunk 6 above)
Comment (webmuniz) [reply src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest04.java:31]: Pascal

File: src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest05.java (hunk 7 above)
Comment (webmuniz) [reply src/academy/devdojo/maratonajava/javacore/streams/test/StreamCollectorsTest05.java:26]: removed

**Review by bryan-cda — APPROVED** [review]

**Pull request merged**